src/source/src/parser/mod.rs

```diff
@@ -33,10 +35,115 @@ mod debezium;
 mod json_parser;
 mod protobuf_parser;
 
-#[derive(Debug, Default)]
-pub struct Event {
-    pub ops: Vec<Op>,
-    pub rows: Vec<Vec<Datum>>,
+/// A builder for building a [`StreamChunk`] from [`SourceColumnDesc`].
+pub struct SourceStreamChunkBuilder {
+    descs: Vec<SourceColumnDesc>,
+    builders: Vec<ArrayBuilderImpl>,
+    op_builder: Vec<Op>,
+}
+
+impl SourceStreamChunkBuilder {
+    pub fn with_capacity(descs: Vec<SourceColumnDesc>, cap: usize) -> Self {
+        let builders = descs
+            .iter()
+            .map(|desc| desc.data_type.create_array_builder(cap))
+            .collect();
+        Self {
+            descs,
+            builders,
+            op_builder: Vec::with_capacity(cap),
+        }
+    }
+
+    pub fn row_writer(&mut self) -> SourceStreamChunkRowWriter<'_> {
+        SourceStreamChunkRowWriter {
+            descs: &self.descs,
+            builders: &mut self.builders,
+            op_builder: &mut self.op_builder,
+        }
+    }
+
+    pub fn finish(self) -> Result<StreamChunk> {
+        Ok(StreamChunk::new(
+            self.op_builder,
+            self.builders
+                .into_iter()
+                .map(|builder| -> Result<_> { Ok(Column::new(Arc::new(builder.finish()?))) })
+                .try_collect()?,
+            None,
+        ))
+    }
+}
+
+/// `SourceStreamChunkRowWriter` is responsible to write one row (Insert/Delete) or two rows
+/// (Update) to the [`StreamChunk`].
+pub struct SourceStreamChunkRowWriter<'a> {
+    descs: &'a [SourceColumnDesc],
+    builders: &'a mut [ArrayBuilderImpl],
+    op_builder: &'a mut Vec<Op>,
+}
+
+/// `WriteGuard` can't be constructed directly in other mods, so it can be used to ensure that the
+/// write methods are called at least once.
+pub struct WriteGuard(());
```

Comment:
Why do we need WriteGuard?

Reply:
> We can ensure every parse call can either write exactly one record or return an Error.

I think it is exactly what `Result<_>` do

Reply:
No, for example, if the function return `Result<()>`, then the parse implementation can be:

```rust
impl SourceParser for DebeziumJsonParser {
    //...
    fn parse(writer: Writer<'_>) -> Result<()> {
        match debezium_op {
            OP_INSERT => writer.insert(""),
            OP_DELETE => writer.delete(""),
            OP_UPDATE => writer.update(""),
            _ => Ok(())
        }
    }
}
```

What happened? we neither write a record nor return an Error.

You may think it's acceptable, but if there are more branches and more complicated logical?

The key difference between `WriteGuard` and `()` is that you can construct a `()` but can't construct a `WriteGuard` in the `parse` implementation.

I want to make the signature more efficient but the same safe as the trivial API: `fn parse(desc, payload) -> Record`, where the `Record` is one of `Insert(Row)`, `Delete(Row)` or `Update(Row, Row)`, so what the trivial signature can do?

1. Ensure that one parse can produce at most one record in the chunk (`Record` but not `Vec<Record>`).
2. Ensure that one parse can produce at least one record in the chunk or error happened (`Result<Record>` but not `Result<Option<Record>>`).

In the high performance signature `fn parse(desc, payload, writer)`, the two properties are still kept:

1. Any of `insert`, `update` and `delete` will consume the ownership of `Writer`, so at most one record was written to the Chunk.
2. You can't construct a `WriteGuard` directly, so you have to call one of `insert`, `update` or `delete` to pass the compilation, so at least one record was written to the Chunk.

Then the signature is as safe as the trivial one but achieve the better performance.

I will try to make the comments more clear.

Reply:
Looks interesting 😃 ... But seriously, the Rust community should give this design pattern a name so that we can reference a link somewhere like a Wikipedia of well-known Rust design patterns...

Reply:
I invented this pattern independently (but I guess not the first one) when I developed TiKV, so I really don’t know the name. But it’s really very useful.